pool/worker.go

```diff
@@ -0,0 +1,250 @@
+		// submitted through this channel may cause a new worker
+		// goroutine to be allocated.
+		requests chan *request
+
+		// work is a channel where new tasks are submitted, but is only
+		// read by active worker gorotuines.
+		work chan *request
+
+		// workerSem is a channel-based sempahore that is used to limit
+		// the total number of worker goroutines to the number
+		// prescribed by the WorkerConfig.
+		workerSem chan struct{}
+
+		wg   sync.WaitGroup
+		quit chan struct{}
+	}
+
+	// request is a tuple of task closure and error channel that is used to
+	// both submit a task to the pool and respond with any errors
+	// encountered during the task's execution.
+	request struct {
+		fn      func(WorkerState) error
+		errChan chan error
+	}
+)
+
+// NewWorker initializes a new Worker pool using the provided WorkerConfig.
+func NewWorker(cfg *WorkerConfig) *Worker {
+	return &Worker{
+		cfg:       cfg,
+		requests:  make(chan *request),
+		workerSem: make(chan struct{}, cfg.NumWorkers),
+		work:      make(chan *request),
+		quit:      make(chan struct{}),
+	}
+}
+
+// Start safely spins up the Worker pool.
+func (w *Worker) Start() error {
+	w.started.Do(func() {
+		w.wg.Add(1)
+		go w.requestHandler()
+	})
+	return nil
+}
+
+// Stop safely shuts down the Worker pool.
+func (w *Worker) Stop() error {
+	w.stopped.Do(func() {
+		close(w.quit)
+		w.wg.Wait()
+	})
+	return nil
+}
+
+// Submit accepts a function closure to the worker pool. The returned error will
+// be either the result of the closure's execution or an ErrWorkerPoolExiting if
+// a shutdown is requested.
+func (w *Worker) Submit(fn func(WorkerState) error) error {
+	req := &request{
+		fn:      fn,
+		errChan: make(chan error, 1),
+	}
+
+	select {
+
+	// Send request to requestHandler, where either a new worker is spawned
+	// or the task will be handed to an existing worker.
+	case w.requests <- req:
+
+	// Fast path directly to existing worker.
+	case w.work <- req:
+
+	case <-w.quit:
+		return ErrWorkerPoolExiting
+	}
+
+	select {
+
+	// Wait for task to be processed.
+	case err := <-req.errChan:
+		return err
+
+	case <-w.quit:
+		return ErrWorkerPoolExiting
+	}
+}
+
+// requestHandler processes incoming tasks by either allocating new worker
+// goroutines to process the incoming tasks, or by feeding a submitted task to
+// an already running worker goroutine.
+func (w *Worker) requestHandler() {
+	defer w.wg.Done()
+
+	for {
+		select {
+		case req := <-w.requests:
+			select {
+
+			// If we have not reached our maximum number of workers,
+			// spawn one to process the submitted request.
+			case w.workerSem <- struct{}{}:
+				w.wg.Add(1)
+				go w.spawnWorker(req)
+
+			// Otherwise, submit the task to any of the active
+			// workers.
+			case w.work <- req:
+
+			case <-w.quit:
+				return
+			}
+
+		case <-w.quit:
+			return
+		}
+	}
+}
+
+// spawnWorker is used when the Worker pool wishes to create a new worker
```

Comment:
👍 

Now that this has been abstracted even more it can be used in other areas of the codebase where we require a pool of works (with a burst limit) to process a series of tasks. With some tweaking, it'll likely also possible to convert the existing `lnwallet.SigPool` to use this new scaffolding.

Reply:
indeed i'm interested to see how the `SigPool` gets adapted :)